src/server/src/main/java/io/cassandrareaper/service/SnapshotManager.java

```diff
@@ -0,0 +1,330 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cassandrareaper.service;
+
+import io.cassandrareaper.AppContext;
+import io.cassandrareaper.ReaperException;
+import io.cassandrareaper.core.Cluster;
+import io.cassandrareaper.core.Node;
+import io.cassandrareaper.core.Snapshot;
+import io.cassandrareaper.core.Snapshot.Builder;
+import io.cassandrareaper.jmx.JmxProxy;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import org.apache.commons.lang3.tuple.Pair;
+import org.joda.time.DateTime;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class SnapshotManager {
+  public static final String SNAPSHOT_PREFIX = "reaper";
+  private static final Logger LOG = LoggerFactory.getLogger(SnapshotManager.class);
+  private final AppContext context;
+  private final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd-HH-mm-ss");
+  private final ExecutorService executor = Executors.newFixedThreadPool(5);
+  private final Cache<String, Snapshot> cache =
+      CacheBuilder.newBuilder()
+          .maximumSize(1000) // Taille Max
+          .expireAfterWrite(1, TimeUnit.HOURS) // TTL
+          .build();
```

Comment:
the only benefit i can see of the cache is that it avoids a call to `context.storage.getSnapshot(..)`.

What's the measured optimisation that this provides? 

Line 173 can be replaced with '.parallelStream()` and the calls to `context.storage.getSnapshot(..)` will execute in parallel. I would think this is more effective (and better on heap) than the cache approach.

Reply:
As I think having the cache will matter in large clusters, I've moved the cache to a `weakValues()` one as discussed together.